### PR TITLE
feat: Manage ratelimits in the REST client

### PR DESCRIPTION
**Pull request by ZeLarpMaster**
The goal of this PR is to make the library follow the requirements detailed in [Discord's HTTP ratelimit documentation](https://discord.com/developers/docs/topics/rate-limits#rate-limits). The only things left to implement later in that documentation page will be detecting when a token becomes invalid and preventing permission issues.

TODO:
- [x] Write `@doc` in `Mobius.Services.RestRatelimiter`
- [x] Tests for `Mobius.Rest.Middleware.Ratelimit`
- [x] Tests for `Mobius.Services.RestRatelimiter`

### CODE REVIEW

**Review by Blond11516 — COMMENTED** [review]
Nothing to say for now! :+1:

File: test/mobius/services/rest_ratelimiter_test.exs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  use ExUnit.Case
+
+  import Mobius.Assertions
+
+  alias Mobius.Services.RestRatelimiter
+
+  describe "wait_ratelimit/1" do
+    test "doesn't wait for global if never updated" do
+      assert_wait_time("global", 5)
+    end
```
Comment (Blond11516) [inline test/mobius/services/rest_ratelimiter_test.exs:11]: So "doesn't wait" == "wait 5 ms"?

File: lib/mobius/services/rest_ratelimiter.ex [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  @doc "Updates a bucket's remaining ratelimit and tracks the route's bucket for later usage"
+  @spec update_ratelimit(any, String.t(), integer, integer) :: :ok
+  def update_ratelimit(route, bucket, remaining, reset_after) do
+    reset = System.monotonic_time(:millisecond) + reset_after
+    GenServer.call(__MODULE__, {:update, route, bucket, remaining, reset})
+  end
```
Comment (Blond11516) [inline lib/mobius/services/rest_ratelimiter.ex]: What would you think of an "update_global_ratelimit" function so we can hide the awkward `update_ratelimit("global", "global", ...)` calls?

File: lib/mobius/services/rest_ratelimiter.ex [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  def handle_call({:request, route}, _from, state) do
+    now = System.monotonic_time(:millisecond)
+
+    case :ets.lookup(__MODULE__, find_bucket(state, route)) do
+      [] -> {:reply, :ok, state}
+      [{_, _, expiration}] when now > expiration -> {:reply, :ok, state}
+      [{_, remaining, _}] when remaining > 0 -> {:reply, :ok, state}
```
Comment (Blond11516) [inline lib/mobius/services/rest_ratelimiter.ex:81]: Shouldn't we check `expiration` and `remaining` together? Should we reply `:ok` if `now > expiration` but `remaining == 0`?

File: test/mobius/rest/middleware/ratelimit_test.exs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    assert_get_time(ctx.client, url, 50)
+  end
+
+  test "waits until reset-after if ratelimit was exceeded", ctx do
+    url = "https://discord.com/api/v6/something"
+    mock(fn %{url: ^url} -> response(-1, 50, 0) end)
+    Tesla.get(ctx.client, url)
+
+    assert_get_time(ctx.client, url, 50)
+  end
```
Comment (Blond11516) [inline test/mobius/rest/middleware/ratelimit_test.exs]: Are those really different? Isn't the edge case `< 0` for remaining anyway?

File: test/mobius/rest/middleware/ratelimit_test.exs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  defp assert_get_time(client, url, expected_time) do
+    client
+    |> Tesla.get(url)
+    |> function_time()
+    |> assert_in_delta(expected_time, 10)
+  end
+
+  defp response(remaining, delay, global_delay) do
```
Comment (Blond11516) [inline test/mobius/rest/middleware/ratelimit_test.exs]: I think this function would gain from taking a keyword list as a parameter instead of 3 distinct values. Right now there's no way to know what the tests are doing without looking at the implementation. Maybe a more descriptive name would help too.

Comment (Blond11516) [reply test/mobius/rest/middleware/ratelimit_test.exs]: More that that, it might be worth splitting it into multiple functions like `response_with_global_ratelimit` and `response_with_route_ratelimit` for example.

In many cases, especially when giving a global delay > 0, you don't actually care what the other values are (expect maybe for "asserting" that they are ignored), so why specify them at all?

File: test/support/assertions.ex [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  @spec function_time(any) :: Macro.output()
+  defmacro function_time(function) do
+    quote do
+      fn -> unquote(function) end
+      |> :timer.tc()
+      |> elem(0)
+      |> :erlang.convert_time_unit(:microsecond, :millisecond)
+    end
+  end
+end
```
Comment (Blond11516) [inline test/support/assertions.ex]: That's not really an assertion...

File: test/mobius/services/rest_ratelimiter_test.exs (hunk 1 above)
Comment (ZeLarpMaster) [reply test/mobius/services/rest_ratelimiter_test.exs:11]: Need some tolerance for the execution time otherwise the test starts to fail randomly. `assert_wait_time` adds a `+- 10ms` to the expected time. So "doesn't wait" is actually "executes in less than 15ms" which is a fair assert imo for a function which would likely wait way more if something wrong were to happen.

File: lib/mobius/services/rest_ratelimiter.ex (hunk 2 above)
Comment (ZeLarpMaster) [reply lib/mobius/services/rest_ratelimiter.ex]: I think that's a pretty good idea, I'll do that

File: lib/mobius/services/rest_ratelimiter.ex (hunk 3 above)
Comment (ZeLarpMaster) [reply lib/mobius/services/rest_ratelimiter.ex:81]: You mean condense both together with an `or`? We could, but imo it's a bit more readable this way as it makes the lines less dense.

File: test/mobius/rest/middleware/ratelimit_test.exs (hunk 4 above)
Comment (ZeLarpMaster) [reply test/mobius/rest/middleware/ratelimit_test.exs]: Oh you're right, I derped while copy pasting stuff :facepalm:

File: test/mobius/rest/middleware/ratelimit_test.exs (hunk 5 above)
Comment (ZeLarpMaster) [reply test/mobius/rest/middleware/ratelimit_test.exs]: That's a good idea, it didn't occur to me how unreadable those tests were after spending so much time on them

File: test/support/assertions.ex (hunk 6 above)
Comment (ZeLarpMaster) [reply test/support/assertions.ex]: That's very true :joy: It was at first, but isn't anymore so I'll move it elsewhere

File: test/mobius/services/rest_ratelimiter_test.exs (hunk 1 above)
Comment (ZeLarpMaster) [reply test/mobius/services/rest_ratelimiter_test.exs:11]: > otherwise the test *could start to fail randomly

This is more likely to happen on the CI since its processing power might be fairly limited. Maybe the tolerance is too high? I'm not sure what I'd change here to make it clearer or better :thinking: Do you have suggestions?

Comment (Blond11516) [reply test/mobius/services/rest_ratelimiter_test.exs:11]: Not really, timing based tests are always kind of a pain so that seems good enough.

File: lib/mobius/services/rest_ratelimiter.ex (hunk 3 above)
Comment (Blond11516) [reply lib/mobius/services/rest_ratelimiter.ex:81]: My original comment wasn't that clear but I'm mostly wondering if the current behaviour is correct or not.

The way the code is currently written, we could let a call through because `now > expiration` even though `remaining <= 0`. Am I correct? If so, is this even possible or do Discord's ratelimiting rules prevent from happening in the first place?

Comment (ZeLarpMaster) [reply lib/mobius/services/rest_ratelimiter.ex:81]: Well when `now > expiration`, the latest `update_route_ratelimit` has expired. This means the bucket's ratelimit has reset and is ready to accept new requests again. So even if `remaining <= 0`, the bucket is actually "full" again at that point in time.

For example, consider the following sequence of events:
- User makes a request
- Discord replies with the HTTP headers `x-ratelimit-remaining: 0` and `x-ratelimit-reset-after: 500`
- Nothing happens for more than 500ms
- User makes another request on the same bucket

At this point in time, the bucket has reset as specified by the `reset-after`. This means the `remaining` resets back to the limit of `5` and the new request will go through, replying with the header `x-ratelimit-remaining: 4` and some reset time in the future.

I hope this makes sense!

Comment (Blond11516) [reply lib/mobius/services/rest_ratelimiter.ex:81]: Yup, makes perfect sense! I just wasn't sure what the exact ratelimiting rules were.

**Review by Blond11516 — APPROVED** [review]
